storage/local/storage.go

```diff
@@ -286,7 +286,47 @@ func (s *memorySeriesStorage) NewIterator(fp clientmodel.Fingerprint) SeriesIter
 		// return any values.
 		return nopSeriesIterator{}
 	}
-	return series.newIterator()
+	return &boundedIterator{
+		it:    series.newIterator(),
+		start: clientmodel.Now().Add(-s.dropAfter),
```

Comment:
An alternative option was to actually check it in the real iterators at call-time. Aside from passing down the retention time all the way this would have resulted in actual differences of possibly many seconds.

At least in `promql/` we are creating all iterators at once in a single AST walk. So the very low millisecond range is what I'd expect at most.
Worth it to keep this transparent, I think.